### PR TITLE
CAP32_WAITBREAK implementation

### PR DESCRIPTION
**Pull request by cpcitor**
This is work in progress towards solving #110.

Code is nice and all, already useful.

It definitely successfully passes these:

<pre>
cap32 -a CAP32_SPEED -a 'mode 1:print"Hello World":call &bd19:call 0' -a CAP32_WAITBREAK -a CAP32_EXIT
cap32 -a CAP32_SPEED -a 'mode 1:print#8,"Hello World":call 0' -a CAP32_WAITBREAK -a CAP32_EXIT
</pre>

It also passes this:

<pre>
cap32 -a CAP32_SPEED -a 'mode 1:print"Hello World":call &bd19:call 0' -a CAP32_WAITBREAK -a CAP32_SCRNSHOT -a CAP32_EXIT
</pre>

This reveals one problem. The spurious Enter keypress after CAP32_WAITBREAK causes a one frame delay which causes the screenshot to miss the bottom half of the screen (because `call 0` had time to start its effect).

<pre>
cap32 -a CAP32_SPEED -a 'mode 1:for a=1 to 22:print"Hello World",a:next:call &bd19:call 0' -a CAP32_WAITBREAK -a CAP32_SCRNSHOT -a CAP32_EXIT
</pre>

# Possible improvements

* Fix SCRNSHOT coming too late by removing spurious Enter keypress.
* Add tests in suite.

What do you think?

### CODE REVIEW

**Comment by cpcitor** [comment]
Looking at why Travis coughs.

**Comment by cpcitor** [comment]
# e2e_test passes

e2e_test on Travis did pass:

https://travis-ci.org/ColinPitrat/caprice32/builds/323654889 says:

> The command "make e2e_test" exited with 0.
> 1.19s$ make clean
> rm -rf obj/ release/ .pc/
> rm -f test_runner test_runner.exe cap32 cap32.exe .debug tags

But the remainder was cancelled because each build took 15-20 minutes. Travis overload?

https://travis-ci.org/ColinPitrat/caprice32/builds/323654889 says:

> The job exceeded the maximum time limit for jobs, and has been terminated.

**Comment by cpcitor** [comment]
# TL;DR: OK to merge

* I continued to work on code, favoring small readable changes.
* Code in this PR passes test.  I believe code quality is good for merge.
* Other improvements discussed below (can be read later).

You can merge if you agree with improvements.

# break-point logic

Code sets `break_point` again even if not useful.  Consequence: every time the CPC resets, the main emulator looped is walked through because `z80_execute` returned `EC_BREAKPOINT`.  Really not a big deal, we're adding one or two walkthrough (in total, not per second) of a loop that gets walked through at least a 50-100 times per second already.  This could be fixed by adding more state like "waiting_for_waitbreak" but I'm doubt the added code would be worth the trouble. What do you think?

# Fixing too long delay between break and snapshot

## Try an expedient first

I tried an expedient avoiding any code change: concatenate the `CAP32_*` strings.

`cap32 -c cap32.cfg -a 'border 13:ink 0,13:ink 1,0:mode 1:for a=1 to 24:print"Hello World",a:next:call &bd19:call 0' -a 'CAP32_WAITBREAKCAP32_SCRNSHOTCAP32_EXIT'`

Not the most readable command line but it allowed to quickly check without spurious linefeed character.

Test still failed: emulator quit before typing the sentence. Why? Because I inserted events from command line in a loop.

## Initial code

Was like this:

``` C++
while { // main emulator loop
   if (event_from_command_line_to_push()) { push_to_SDL(); }
   while (SDL_PollEvent(&event)) { handle_SDL_event(); }
   if (!CPC.paused) { emulate_some_more(); } else { sleep(); }
}
```

The first `if` is best option when `event_from_command_line_to_push` must not fire too quickly, like key presses.

## Inserting many events makes interesting future use cases

I changed the `if` into a `while`...

``` C++
while { // main emulator loop
   while (event_from_command_line_to_push()) { push_to_SDL(); }  /*<-- This line*/
   while event_from_SDL { handle_SDL_event(); }
   if (!CPC.paused) { emulate_some_more(); } else { sleep(); }
}
```

... and now understand its actual use case.  It could allow to handle several events at the exact emulated time, for example screenshot and snapshot.

Here are use cases:

* Cycle-accurate integrated tests of the emulator (allowing to test down both to the Z80 register and screen content reliably)

`cap32 -c cap32.cfg -a 'run"captest' --break-address $ADDRESS1 -a 'CAP32_WAITBREAKCAP32_SCRNSHOTCAP32_SNAPSHOT' --break-address $ADDRESS2 -a 'CAP32_WAITBREAKCAP32_SCRNSHOTCAP32_SNAPSHOT' ...`

and diff in `test.sh` compares a whole output directory (containing `printer.dat`, snapshots, screenshots) like this:

    if diff -ur model/ output/ ...

* Cycle-accurate tests of cpc-level code (I'd like this)

`cap32 -c cap32.cfg -a 'run"cpccode.tst' --break-address $ADDRESS1 -a 'CAP32_WAITBREAKCAP32_SCRNSHOTCAP32_SNAPSHOT' --break-address $ADDRESS2 -a 'CAP32_WAITBREAKCAP32_SCRNSHOTCAP32_SNAPSHOT' ...`

and diff in `test.sh` compares a whole output directory (containing `printer.dat`, snapshots, screenshots) like this:

    if diff -ur model/ output/ ...


In other words, this would be reliable:

`cap32 ... -a 'CAP32_WAITBREAKCAP32_SCRNSHOT`

this would not be reliable:

`cap32 ... -a 'CAP32_WAITBREAKCAP32_SCRNSHOTCAP32_SNAPSHOT'`

**Without this "while" loop, multiple-event tests can pass one day and fail another because even if everything is deterministic in the emulated CPC, some external circumstance or code change altered how many cycles `z80_execute()` runs before it returns and let next event from command line be inserted.**

## Okay, but I only want to do break+screenshot for now.

Possible solutions for the eager loop:

1. change the `while` back into a `if`(to `SDL_PushEvent`): works again, but then there will be an implementation-dependent delay between any `SDL_PushEvent`, breaking cycle-accurate handling of multiple events.
2. keep the loop, set a specific condition to break the loop when event is `CAP32_WAITBREAK`.  Somehow more rigorous but makes less clear code to understand.

Also notice, that:

* command line may cause a loop to insert lots of SDL events.
* synthesizing and pushing SDL events always feels a little hackish yet remains somehow natural when we really mean keypresses.  WAITBREAK is not a keypress

A different code structure would not even need to *insert* SDL events or even to *create* some.  Solutions sketched below.

### Changing overall loop

All this reminds that `CAP32_*` from `enum CAP32_KEYS` really represent human-triggered keypresses: precise timing doesn't matter, the toggle issue ("Which state am I in?") is arbitrated by visual OSD feedback.

Events from command line would better be served by independent data structure but this implies changing the loops.

Turning this:

``` C++
while // main emulator loop
  if (event_from_command_line_to_push()) { push_to_SDL(); }
  while (SDL_PollEvent(&event)) { handle_SDL_event(); }
  if (!CPC.paused) { emulate_some_more(); } else { sleep(); }
}

```

#### Don't push `SDL_Event`

Into (3):

``` C++
while { // main emulator loop
   do { // start event insertion loop
      if (!SDL_PollEvent(&event)) { // no SDL event, see if scripted event
         if (!event_from_script(&event)) {
            break; // break event insertion loop
         }
      }
      handle_SDL_event(); // There, CAP32_WAITBREAK would also tell to break the loop.
   }
   while (true); // end event insertion loop

   if (!CPC.paused) { emulate_some_more(); } else { sleep(); }
}

```

Pros/cons:

* + allows to handle `CAP32_WAITBREAK` in only one place, no wart
* + we can handle many events before ticking emulation again
* - events are still `SDL_Events`
* - somehow convoluted code/loop logic
* - any keypress by actual user will be inserted in the middle of the scripted keypresses (that alone could be fixed but read on)

#### Handle scripted events like they are

Or even (4):

Implement scripted events completely separately from key down/up events.

``` C++
while { // main emulator loop
   while (event_from_script()) {
      handle_scripted_event();
      /* This handles emulated keypresses more naturally,
         ascii->CPC_keypress, no double conversion via SDL_KeyEvent.
         Also `{CAP32_CMD_WAITBREAK}` cleanly, on wart.
      */
   }
   while (SDL_PollEvent(&event)) {
      handle_SDL_event();
      /* including key press/release event,
         including `CAP32_KEYS` when actually pressed from host keyboard;*/
   }
   if (!CPC.paused) { emulate_some_more(); } else { sleep(); }
}
```

Pros/cons:

* + allows to handle `CAP32_WAITBREAK` in only one place, no wart
* + we can handle many events if needed before ticking emulation again
* + clean code/loop logic
* + will allow to simplify some existing logic in other source files
    * no more "CAP32_WAITBREAK  SDLK_BREAK" in language keymaps
    * no more double conversion SDK_Keys -> CAP -> ...
* + clean design open to all sort of instrumentation and extensions, no SDL-tied hacks
    * fine-grained CPC keypresses like Ctrl-KeyDown, Enter-KeyDown, Enter-KeyUp, Ctrl-KeyUp.
    * `{CAP32_CMD_SCRNSHOT}` different from the keypress because reproducible file names,
    * `{CAP32_CMD_SAVESNA}` different from the keypress because reproducible file names,


Another topic: it's best to make scripted events robust in case of interacting user typing something before script is finished.  One solution is ignoring regular keypress `SDL_Event`, which is easier to do when they are not mixed with scripted events.


# Conclusion

Sketched 4 possible next steps to improve.

1. very easy, solves local problem, fully YAGNI-approved
2. somehow convoluted
3. somehow convoluted
4. open but more work. YAGNI suggests: best but later.

So far I have:

* implemented path 1 and 
* added `test/integrated/scripted_operations_timing` which passes on my machine and on Travis
* think option 4 is best but waiting for (1) actual need (2) your feedback 

Do you OK the PR? What do you think of the evolution paths?

Thank you for your attention.

**Comment by ColinPitrat** [comment]
I agree it's not worth avoiding the additional main loop in case of breakpoint.

I agree Option 4 is better. It would change autocmd from a quick&dirty hack to a fully-fledged feature.
May be a lot of work though.
